crates/cranelift/src/func_environ/stack_switching/instructions.rs

```diff
@@ -302,8 +304,16 @@ pub(crate) mod stack_switching_helpers {
             required_capacity: u32,
             existing_slot: Option<StackSlot>,
         ) -> StackSlot {
-            let align = u8::try_from(std::mem::align_of::<T>()).unwrap();
-            let entry_size = u32::try_from(std::mem::size_of::<T>()).unwrap();
+            // TODO: hack around host pointer size being mixed up with target
+            let (align, entry_size) =
+                if core::any::TypeId::of::<T>() == core::any::TypeId::of::<*mut u8>() {
+                    (env.pointer_type().bytes() as u8, env.pointer_type().bytes())
+                } else {
+                    (
+                        u8::try_from(std::mem::align_of::<T>()).unwrap(),
+                        u32::try_from(std::mem::size_of::<T>()).unwrap(),
+                    )
+                };
```

Comment:
This stuff should be a method on the `PtrSize` trait that we define in `vmoffsets.rs` (maybe `fn vm_host_array_layout(&self) -> (u8, u32)`) and ultimately we should have access to that method here via `env.offsets.ptr`.

We should definitely avoid any kind of `std::mem::{size,align}_of` calls, as that ends up being pretty fragile (doesn't work for `f64` across 32- and 64-bit architectures, for example, where the alignment is different despite size being the same).

Reply:
Makes sense, I'll take a look at making changes to revise VMHostArray and any other problematic parts I can find (or see if @dhil can help out).  I wanted to get a basic version of the commit up to get this kind of feedback as doing operations on the generic T here (from the original impl) seemed problematic for the reasons you describe.